### PR TITLE
added object detection with vision transformer py file

### PR DESCRIPTION
**Pull request by kdave123**
This example implements object detection with vision transformer. Overall code is ~300 lines excluding comments. Have used some code from existing examples on vision transformers available on keras.io. Please take a look, thanks.

### CODE REVIEW

**Comment by kdave123** [comment]
@fchollet hi,  can you please review it? or is there anything pending on my end please let me know. Thanks

File: examples/vision/object_detection_using_vision_transformer.py [hunk 1]
```diff
@@ -0,0 +1,605 @@
...
+Last modified: 2022/03/27
+Description: A simple keras implementation for object detection using transformers.
+"""
+
+
+"""
+## Introduction
+
+ This example is a an implementation of the [Vision Transformer (ViT)]
+ (https://arxiv.org/abs/2010.11929)
```
Comment (fchollet) [inline examples/vision/object_detection_using_vision_transformer.py:14]: Format mkdown links on a single line each. Make sure to read and follow instructions from the [contributor guide](https://github.com/keras-team/keras-io/blob/master/contributor_guide.md).

File: examples/vision/object_detection_using_vision_transformer.py [hunk 2]
```diff
@@ -0,0 +1,605 @@
...
+from tensorflow import keras
+from tensorflow.keras import layers
+import tensorflow_addons as tfa
+import matplotlib.pyplot as plt
+import numpy as np
+import cv2
+import os
+import scipy.io
+
+from tensorflow.keras.preprocessing.image import img_to_array
```
Comment (fchollet) [inline examples/vision/object_detection_using_vision_transformer.py:41]: Access these from `keras.utils.img_to_array` and `keras.utils.load_img`

File: examples/vision/object_detection_using_vision_transformer.py [hunk 3]
```diff
@@ -0,0 +1,605 @@
...
+from tensorflow.keras.preprocessing.image import load_img
+
+import tensorflow as tf
+import datetime
+
+"""
+ To run the TensorBoard from a Jupyter notebook,
+ load the TensorBoard notebook extension inside notebook
+```
+%load_ext tensorboard
```
Comment (fchollet) [inline examples/vision/object_detection_using_vision_transformer.py:51]: This should be moved to the section where it is used.

File: examples/vision/object_detection_using_vision_transformer.py [hunk 4]
```diff
@@ -0,0 +1,605 @@
...
+
+# Dataset
+# http://www.vision.caltech.edu/Image_Datasets/Caltech101/
+# Download the images '101_ObjectCategories.tar.gz' and
+# annotations 'Annotations.tar' available seperately
+# Extract images (using airplanes here) to images folder and
+# all annotations to another folder.
+
+# Path to images and annotations
+PATH_IMAGES = "C:/Users/DAVE/Desktop/Object detection with transformers/airplanes/"
```
Comment (fchollet) [inline examples/vision/object_detection_using_vision_transformer.py:67]: Rather than hard coding paths on your own disk, write Python code that downloads the file (e.g. using `keras.utils.get_file`, uncompressed the data, etc.

File: examples/vision/object_detection_using_vision_transformer.py [hunk 5]
```diff
@@ -0,0 +1,605 @@
...
+# Some Code referred from
+# https://keras.io/examples/vision/image_classification_with_vision_transformer/
+
+image_size = 64  # resize input images to this size
+patch_size = 8  # Size of the patches to be extracted from the input images
+
+data_augmentation = keras.Sequential(
+    [
+        layers.Normalization(),
+        layers.Resizing(image_size, image_size),
```
Comment (fchollet) [inline examples/vision/object_detection_using_vision_transformer.py:145]: Resizing is already done on line 113.

File: examples/vision/object_detection_using_vision_transformer.py [hunk 6]
```diff
@@ -0,0 +1,605 @@
...
+
+# Some Code referred from
+# https://keras.io/examples/vision/image_classification_with_vision_transformer/
+
+image_size = 64  # resize input images to this size
+patch_size = 8  # Size of the patches to be extracted from the input images
+
+data_augmentation = keras.Sequential(
+    [
+        layers.Normalization(),
```
Comment (fchollet) [inline examples/vision/object_detection_using_vision_transformer.py:144]: Normalization or rescaling should happen as part of the model, not as part of the augmentation stage.

File: examples/vision/object_detection_using_vision_transformer.py [hunk 7]
```diff
@@ -0,0 +1,605 @@
...
+## Implement patch creation layer
+"""
+
+
+class Patches(layers.Layer):
+    def __init__(self, patch_size):
+        super(Patches, self).__init__()
+        self.patch_size = patch_size
+
+    #     Override function to avoid error while saving model
```
Comment (fchollet) [inline examples/vision/object_detection_using_vision_transformer.py:177]: Include it or don't include it, but don't leave it commented-out.

File: examples/vision/object_detection_using_vision_transformer.py [hunk 8]
```diff
@@ -0,0 +1,605 @@
...
+plt.axis("off")
+
+resized_image = tf.image.resize(
+    tf.convert_to_tensor([image]), size=(image_size, image_size)
+)
+patches = Patches(patch_size)(resized_image)
+# print(f"Image size: {image_size} X {image_size}")
+# print(f"Patch size: {patch_size} X {patch_size}")
+# print(f"Patches per image: {patches.shape[1]}")
+# print(f"Elements per patch: {patches.shape[-1]}")
```
Comment (fchollet) [inline examples/vision/object_detection_using_vision_transformer.py:227]: Likewise.

File: examples/vision/object_detection_using_vision_transformer.py [hunk 9]
```diff
@@ -0,0 +1,605 @@
...
+class PatchEncoder(layers.Layer):
+    def __init__(self, num_patches, projection_dim):
+        super(PatchEncoder, self).__init__()
+        self.num_patches = num_patches
+        self.projection = layers.Dense(units=projection_dim)
+        self.position_embedding = layers.Embedding(
+            input_dim=num_patches, output_dim=projection_dim
+        )
+
+    #     Override function to avoid error while saving model
```
Comment (fchollet) [inline examples/vision/object_detection_using_vision_transformer.py:254]: Likewise.

File: examples/vision/object_detection_using_vision_transformer.py [hunk 10]
```diff
@@ -0,0 +1,605 @@
...
+
+    # return Keras model.
+    return keras.Model(inputs=inputs, outputs=bounding_box)
+
+
+"""
+## Run experiment
+"""
+
+import keras.backend as K
```
Comment (fchollet) [inline examples/vision/object_detection_using_vision_transformer.py:345]: This is not used.

File: examples/vision/object_detection_using_vision_transformer.py [hunk 11]
```diff
@@ -0,0 +1,605 @@
...
+from keras.applications.vgg16 import VGG16
+from keras.models import Model
+
+# import pandas as pd
+
+tf.config.run_functions_eagerly(True)
```
Comment (fchollet) [inline examples/vision/object_detection_using_vision_transformer.py:351]: Why?

File: examples/vision/object_detection_using_vision_transformer.py [hunk 12]
```diff
@@ -0,0 +1,605 @@
...
+    # return Keras model.
+    return keras.Model(inputs=inputs, outputs=bounding_box)
+
+
+"""
+## Run experiment
+"""
+
+import keras.backend as K
+from keras.applications.vgg16 import VGG16
```
Comment (fchollet) [inline examples/vision/object_detection_using_vision_transformer.py:346]: Why use VGG16? Almost any other model would do better.

File: examples/vision/object_detection_using_vision_transformer.py [hunk 13]
```diff
@@ -0,0 +1,605 @@
...
+from keras.models import Model
```
Comment (fchollet) [inline examples/vision/object_detection_using_vision_transformer.py:347]: Just access it as `keras.Model`.

File: examples/vision/object_detection_using_vision_transformer.py [hunk 14]
```diff
@@ -0,0 +1,605 @@
...
+    model.compile(optimizer=optimizer, loss=keras.losses.MeanSquaredError())
+
+    checkpoint_filepath = "logs/"
+    checkpoint_callback = keras.callbacks.ModelCheckpoint(
+        checkpoint_filepath,
+        monitor="val_loss",
+        save_best_only=True,
+        save_weights_only=True,
+    )
+    #    log_dir = (
```
Comment (fchollet) [inline examples/vision/object_detection_using_vision_transformer.py:369]: Include it or don't, but don't leave it commented-out.

**Review by fchollet — CHANGES_REQUESTED** [review]
Thanks for the PR! Here's some initial feedback.

**Comment by kdave123** [comment]
@fchollet Thank you for reviewing, I will make the requested changes

**Comment by fchollet** [comment]
@[kdave123](https://github.com/kdave123) great, thanks! By the way, do you have someone who could help you improve the text copy?

**Comment by kdave123** [comment]
@fchollet Yes, I will work on it. Also, if you know anyone/anywhere specific who can help with it then please let me know. Thanks.

File: examples/vision/object_detection_using_vision_transformer.py (hunk 12 above)
Comment (kdave123) [reply examples/vision/object_detection_using_vision_transformer.py]: unsed code, havent used VGG

**Comment by kdave123** [comment]
@fchollet I have made the changes. Please take a look!

File: examples/vision/object_detection_using_vision_transformer.py [hunk 15]
```diff
@@ -0,0 +1,499 @@
...
+This is an implementation of the
+[Vision Transformer (ViT)](https://arxiv.org/abs/2010.11929)
+model by Alexey Dosovitskiy et al. applied to the object detection task.
+The model is trained on
+[Caltech 101 dataset](http://www.vision.caltech.edu/datasets/),
+to detect an airplane in the given image. The article demonstrates
+that a pure transformer applied directly to sequences of the image
+patches can perform well on object detection tasks.
+
+ This example requires TensorFlow 2.4 or higher, and
```
Comment (fchollet) [inline examples/vision/object_detection_using_vision_transformer.py:21]: Remove indent for this paragraph

Comment (kdave123) [reply examples/vision/object_detection_using_vision_transformer.py]: Done

File: examples/vision/object_detection_using_vision_transformer.py [hunk 16]
```diff
@@ -0,0 +1,499 @@
...
+image_size = 224  # resize input images to this size
+
+images, targets = [], []
+
+# loop over the annotations and images, preprocess them and store in lists
+for i in range(0, len(annot_paths)):
+    # Access bounding box coordinates
+    annot = scipy.io.loadmat(PATH_ANNOT + annot_paths[i])["box_coord"][0]
+
+    topLeft_X = annot[2]
```
Comment (fchollet) [inline examples/vision/object_detection_using_vision_transformer.py:87]: Always use sake case for variable names

File: examples/vision/object_detection_using_vision_transformer.py [hunk 17]
```diff
@@ -0,0 +1,499 @@
...
+ pip install -U tensorflow-addons
+```
+"""
+
+"""
+## Imports and setup
+"""
+
+import numpy as np
+import tensorflow as tf
```
Comment (fchollet) [inline examples/vision/object_detection_using_vision_transformer.py:36]: Also `from tensorflow import keras` and `from tensorflow.keras import layers` since these 2 are commonly used modules in this file.

File: examples/vision/object_detection_using_vision_transformer.py [hunk 18]
```diff
@@ -0,0 +1,499 @@
...
+        positions = tf.range(start=0, limit=self.num_patches, delta=1)
+        return self.projection(patch) + self.position_embedding(
+            positions
+        )  # return encoded
+
+
+"""
+## Build the ViT model
+
+ The ViT model has multiple Transformer blocks,
```
Comment (fchollet) [inline examples/vision/object_detection_using_vision_transformer.py:248]: Almost remove paragraph indent here

File: examples/vision/object_detection_using_vision_transformer.py [hunk 19]
```diff
@@ -0,0 +1,499 @@
...
+    iou = interArea / float(boxAArea + boxBArea - interArea)
+
+    # return ioU
+    return iou
+
+
+i, mIoU = 0, 0
+
+# Compare results for 10 images in the test set
+for inputImage in x_test[:10]:
```
Comment (fchollet) [inline examples/vision/object_detection_using_vision_transformer.py:415]: Use snake case

File: examples/vision/object_detection_using_vision_transformer.py [hunk 20]
```diff
@@ -0,0 +1,499 @@
...
+    # calculate intersection over union by taking intersection
+    # area and dividing it by the sum of predicted bb + ground-truth
+    # bb areas - the interesection area
+    iou = interArea / float(boxAArea + boxBArea - interArea)
+
+    # return ioU
+    return iou
+
+
+i, mIoU = 0, 0
```
Comment (fchollet) [inline examples/vision/object_detection_using_vision_transformer.py:412]: Use snake case

File: examples/vision/object_detection_using_vision_transformer.py [hunk 21]
```diff
@@ -0,0 +1,499 @@
...
+
+import matplotlib.patches as patches
+
+# Saves the model in current path
+vit_object_detector.save("vit_object_detector.h5", save_format="h5")
+
+# To calculate IoU (intersection over union, given two bounding boxes)
+def bounding_box_intersection_over_union(boxA, boxB):
+    # get (x, y) coordinates of intersection of bounding boxes
+    xA = max(boxA[0], boxB[0])
```
Comment (fchollet) [inline examples/vision/object_detection_using_vision_transformer.py:391]: Use snake case

**Review by fchollet — CHANGES_REQUESTED** [review]
Thanks for the update! It looks very nice now. I just have some style nits. Once addressed I'll do a final round of copyedits before we publish. Thank you!

File: examples/vision/object_detection_using_vision_transformer.py (hunk 21 above)
Comment (kdave123) [reply examples/vision/object_detection_using_vision_transformer.py]: done

**Comment by kdave123** [comment]
@fchollet I have pushed the updates. The requested changes are addressed.

**Comment by kdave123** [comment]
@fchollet formatted black with a newer version, shouldn't be fail now

**Review by fchollet — COMMENTED** [review]
Thanks for the update! I applied the final copyedits. Please pull the changes and add the generated files 👍

**Comment by kdave123** [comment]
I will do the needful, Thank you @fchollet

**Comment by kdave123** [comment]
@fchollet I have added the generated files, can you please take a look?

**Review by fchollet — APPROVED** [review]
Awesome! I'll take it from there. Thank you for the great contribution!

**Comment by kdave123** [comment]
@fchollet It was great working with you, thank you for your support and active collaboration. Looking forward to working with you in the future.
